neqo-helper/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+            IpDontFrag, IpRecvTos, IpRecvTtl, IpTos, IpTtl, Ipv6DontFrag, Ipv6HopLimit,
+            Ipv6RecvHopLimit, Ipv6RecvTClass, Ipv6TClass,
+        },
+        AddressFamily, ControlMessage, ControlMessageOwned, MsgFlags, SockaddrLike,
+        SockaddrStorage,
+    },
+};
+
+use neqo_common::Datagram;
+
+// Bind a UDP socket and set some default socket options.
+pub fn bind(local_addr: SocketAddr) -> io::Result<UdpSocket> {
+    let socket = match UdpSocket::bind(local_addr) {
+        Err(e) => {
+            eprintln!("Unable to bind UDP socket: {}", e);
+            exit(1)
+        }
+        Ok(s) => {
+            // Don't let the host stack or network path fragment our IP packets
+            // (RFC9000, Section 14).
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpDontFrag, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6DontFrag, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 type-of-service (TOS) and IPv6 traffic class
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTos, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvTClass, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 time-to-live (TTL) and IPv6 hop count
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTtl, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvHopLimit, &true),
+            };
+            assert!(res.is_ok());
+            s
+        }
+    };
+    Ok(socket)
+}
+
+fn to_sockaddr(addr: SocketAddr) -> SockaddrStorage {
+    SockaddrStorage::from(addr)
+}
+
+pub fn emit_datagram(fd: i32, d: Datagram) -> io::Result<()> {
+    let iov = [IoSlice::new(&d[..])];
+    let tos = d.tos() as i32;
+    let ttl = d.ttl() as i32;
+    let cmsg_tos = match d.destination() {
+        SocketAddr::V4(..) => ControlMessage::IpTos(&tos),
+        SocketAddr::V6(..) => ControlMessage::Ipv6TClass(&tos),
+    };
+    let cmsg_ttl = match d.destination() {
+        SocketAddr::V4(..) => ControlMessage::IpTtl(&ttl),
+        SocketAddr::V6(..) => ControlMessage::Ipv6HopLimit(&ttl),
+    };
+    let sent = sendmsg(
+        fd,
+        &iov,
+        &[cmsg_tos, cmsg_ttl],
+        MsgFlags::empty(),
+        Some(&to_sockaddr(d.destination())),
+    )
+    .unwrap();
+    if sent != d.len() {
+        eprintln!("Unable to send all {} bytes of datagram", d.len());
+    }
+    Ok(())
+}
+
+fn to_socket_addr(addr: &SockaddrStorage) -> SocketAddr {
+    let fam = addr.family().unwrap();
+
+    let port = match fam {
+        AddressFamily::Inet => addr.as_sockaddr_in().unwrap().port(),
+        AddressFamily::Inet6 => addr.as_sockaddr_in6().unwrap().port(),
+        _ => todo!(),
```

Comment:
```suggestion
        _ => unreachable!(),
```

I think that we're safe with that in this case.  IPv10 is still a ways off.